### DISCUSSION TITLE
How to install knu/zsh-manydots-magic

### DISCUSSION THREAD

**Discussion by protist** [zimfw/zimfw#481] [Q&A] [answered]
I've been struggling with this for a while, although I feel it should be simple. Sorry if this is stupid question!

How do I install [knu/zsh-manydots-magic](https://github.com/knu/zsh-manydots-magic)?

I've tried permutations of the following in `~/.zimrc` (individually) then `zimfw install`, but all fail
```
zmodule knu/zsh-manydots-magic
zmodule /path/to/zsh-manydots-magic/manydots-magic
zmodule knu/zsh-manydots-magic --autoload manydots-magic --cmd manydots-magic
```

I can see the GitHub repo was downloaded to `~/.zim/modules/zsh-manydots-magic`.

It does work if I manually move the `manydots-magic` file into a directory from `$fpath`, and put the commands as per the repo into `~/.zshrc`:
```
autoload -Uz manydots-magic
manydots-magic
```

What is the "zim-way" to install this?

**Comment by ericbn** [comment, marked as answer]
Hi @protist.

Looking at [knu/zsh-manydots-magic's how-to-set-up instructions](https://github.com/knu/zsh-manydots-magic#how-to-set-up), it mentions:
* "Put the file `manydots-magic` somewhere in your `$fpath`"
* `autoload -Uz manydots-magic`
* `manydots-magic`

This translates to:
`zmodule knu/zsh-manydots-magic --fpath . --autoload manydots-magic --cmd manydots-magic`

I've also faced the https://github.com/knu/zsh-manydots-magic/issues/2 issue, regardless of initializing the module with Zim or manually adding the initialization code to my .zshrc file. Had to disable zsh-users/zsh-autosuggestions to make it work.

**Reply by protist** [reply]
Thanks @ericbn! That works perfectly in terms of installing.

In terms of the issue you linked, I tried installing [twang817's fork](https://github.com/twang817/zsh-manydots-magic) mentioned in the issue instead. I manually deleted `~/.zim/modules/zsh-manydots-magic`, changed the `zmodule` reference, and ran `zimfw install`. I confirmed the new module was indeed the correct one (with `git remote show origin`).

However in both cases backspace is still broken, albeit with a new error: `_zsh_highlight_call_widget: maximum nested function level reached`. The fork works fine on my other system where I install it externally, and the upstream version does fail with a different error message, so I'm not sure if this is relevant to Zim or something else.

**Reply by ericbn** [reply]
FYI, the double-dot-expand feature in Zim's input module does the minimal of what zsh-manydoes-magic does, but in 13 lines of code.

This works:
* `ruby -e '(1...`
* `git log branch...`

But this does not work:
* `git diff ...branch`
* the backspace key automatically deleting the last group

**Reply by ericbn** [reply]
This worked for me with zsh-users/zsh-autosuggestions:
```
zmodule twang817/zsh-manydots-magic --name twang817-manydots-magic --fpath . --autoload manydots-magic --cmd manydots-magic
```
(I'm just using the name to install this in a separate directory)

**Reply by protist** [reply]
Thanks for all that info. Unfortunately I still can't get it working. It's not my server and it has zsh 5.2 installed, so I tried to run zsh 5.8 instead. That still came up with an error, but slightly different.
```
_zsh_highlight_call_widget: maximum nested function level reached; increase FUNCNEST?
```

Since the "manual install" version works for both version (as per the original post), I suspect that it's still possibly something to do with Zim?

**Reply by ericbn** [reply]
Can you share your complete .z{shenv,profile,shrc,login,logout,imrc} files? Maybe we can try to reproduce it in an isolated Docker container?

**Reply by protist** [reply]
Sorry for the delay. I spent some time troubleshooting this, trying to get a minimal failing config. It turns out that I had this error because I had remnants of an older zim install in `~/.zshrc`, which was sourcing `zim.zsh init -q` and `init.zsh` twice! Once I removed the duplicate commands, it worked perfectly.

Thanks for all your help!
